poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+# Poseidon sponge construction
+
+- Version: 0.0.0
+- Maintainer: Jack Grigg
+
+## Introduction
+
+This document specifies the Poseidon sponge construction, a duplex sponge [[BDPV2011]]
+instantiated with the [Poseidon permutation]. It also specifies several instantiations of
+the sponge for use in various hashing domains.
+
+[BDPV2011]: https://keccak.team/files/SpongeDuplex.pdf
+[Poseidon permutation]: ./poseidon-permutation.md
+
+## Notation and conventions
+
+- `[a, b)` means the sequence of integers from `a` inclusive to `b` exclusive.
+- `[x; N]` means an array of length `N` where every entry is filled with `x`.
+- `y[i]` means the element of the array `y` at index `i`.
+- `len(y)` means the length of the array `y`. For example, `len([x; N]) == N`.
+
+## Duplex sponge
+
+The duplex sponge `PoseidonSponge` is built around `PoseidonPermutation`, a permutation of
+width `T`, which encapsulates the specific Poseidon parameters for the sponge. It has a
+parameter `RATE`, the number of input field elements processed by each invocation of the
+permutation, which is at most `T - 1`. The value `C = T - RATE` is called the capacity.
+
+> TODO: Describe how the permutation instance and the rate relate to each other and to the
+> desired security parameter `M`.
+
+`PoseidonSponge` has several internal properties:
+- An internal state of `T` field elements.
+- A mode variable that is either `absorbing` or `squeezing`.
+- A cache of field elements corresponding to the current mode.
```

Comment:
Expand on this bullet point, to make it clear that this cache will contain either unmerged absorbed values, or unsqueezed output values.

Reply:
This does not match how the sponge is currently specified. The cache is exactly that: a cache of elements that have either not yet been absorbed into the state, or have already been extracted from the state. An index into the `RATE` portion of the state would only be able to replace the latter part (squeezing elements) in the current spec; to replace the absorbing parts, we'd need to change the duplex abstraction to not be single-pass, and instead incrementally add the input to the state.